MoreLinq/Traverse.cs

```diff
@@ -0,0 +1,97 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Felipe Sateler. All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Generates a sequence by traversing an element according to the <paramref name="nextSteps"/> generator function. 
+        /// The returned elements are traversed Breadth First.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="seed">The value of the first element in the sequence</param>
+        /// <param name="nextSteps">The function that produces the children of each element</param>
+        /// <returns>A sequence containing the generated values</returns>
+        /// <remarks>
+        /// This function defers element generation until needed and streams the results.
+        /// The progression function is not checked for loops. If you need the resulting sequence to be finite
+        /// you must ensure that loops are not produced by <paramref name="nextSteps"/>
+        /// </remarks>
+        public static IEnumerable<T> TraverseBreadthFirst<T>(T seed, Func<T, IEnumerable<T>> nextSteps)
+        {
+            if (nextSteps == null) {
+                throw new ArgumentNullException("nextSteps");
+            }
+            return TraverseBreadthFirstImpl(seed, nextSteps);
+        }
+
+        private static IEnumerable<T> TraverseBreadthFirstImpl<T>(T seed, Func<T, IEnumerable<T>> nextSteps)
+        {
+            var queue = new Queue<T>();
+            queue.Enqueue(seed);
+
+            while (queue.Count != 0) {
+                var cur = queue.Dequeue();
+                yield return cur;
+                foreach (var child in nextSteps(cur)) {
```

Comment:
If you do go with BSD-style brace layout then you can skip it here for single-statement `for` loop.